packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/component_builder.rs

```diff
@@ -0,0 +1,95 @@
+//! A `ComponentBuilder` turns DAST components into fully hydrated components that are ready for use in Core.
+//! Unlike DAST, which is schema agnostic, components represent element types that are actually implemented in DoenetML.
+
+use std::{collections::HashMap, str::FromStr};
+
+use crate::{
+    components::{ComponentAttributes, ComponentNode},
+    dast::flat_dast::{NormalizedNode, NormalizedRoot, UntaggedContent},
+    ComponentPointerTextOrMacro, ExtendSource,
+};
+
+use super::{ComponentEnum, _error::_Error, _external::_External};
+
+pub struct ComponentBuilder {
+    /// Hydrated components that are ready for use in Core.
+    pub components: Vec<ComponentEnum>,
+}
+
+impl Default for ComponentBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ComponentBuilder {
+    pub fn new() -> Self {
+        Self {
+            components: Vec::new(),
+        }
+    }
+
+    pub fn from_normalized_root(normalized_root: &NormalizedRoot) -> Self {
+        let components = normalized_root
+            .nodes
+            .iter()
+            .enumerate()
+            .map(|(idx, node)| match node {
+                NormalizedNode::Element(elm) => {
+                    let mut component = ComponentEnum::from_str(&elm.name).unwrap_or_else(|_| {
+                        // If we didn't find a match, then create a component of type external
+                        ComponentEnum::_External(_External {
+                            name: elm.name.clone(),
+                            ..Default::default()
+                        })
+                    });
+                    // XXX: we temporarily fill each required attribute with an empty vector.
+                    // This will be removed when typed attributes are integrated.
+                    let attributes: HashMap<&'static str, _> = HashMap::from_iter(
+                        component
+                            .get_attribute_names()
+                            .iter()
+                            .map(|&name| (name, Vec::new())),
+                    );
+                    component.initialize(
+                        elm.idx,
+                        elm.parent,
+                        elm.extending
+                            .clone()
+                            .map(|ref_resolution| ExtendSource::Component(ref_resolution.idx())),
```

Comment:
`ExtendSource::Component` is the correct thing to create if there is no remaining path. However, if there is a remaining path, then you'll need to create an `ExtendSource::StateVar`, assuming there is a match to the remaining path. If there isn't a match, then the component should be replaced with an error.